mlrun/runtimes/pod.py

```diff
@@ -411,6 +420,264 @@ def enrich_resources_with_default_pod_resources(
+                new_node_selector_terms = self._get_node_selector_terms_without_provided_node_selector_requirements(
+                    node_selector_terms=node_selector.node_selector_terms,
+                    node_selector_requirements_to_remove=node_selector_requirements,
+                )
+                # check whether there are node selector terms to add to the new list of required terms
+                if len(new_node_selector_terms) > 0:
+                    new_required_during_scheduling_ignored_during_execution = (
+                        client.V1NodeSelector(
+                            node_selector_terms=new_node_selector_terms
+                        )
+                    )
+            # check if there is something to set into required_during_scheduling_ignored_during_execution
+            if (
+                not node_affinity.preferred_during_scheduling_ignored_during_execution
+                and not new_required_during_scheduling_ignored_during_execution
+            ):
+                self.affinity.node_affinity.required_during_scheduling_ignored_during_execution = (
+                    None
+                )
+                return
+
+            self._initialize_affinity()
+            self._initialize_node_affinity()
+            self.affinity.node_affinity.required_during_scheduling_ignored_during_execution = (
+                new_required_during_scheduling_ignored_during_execution
+            )
+
+    @staticmethod
+    def _get_node_selector_terms_without_provided_node_selector_requirements(
+        node_selector_terms: typing.List[client.V1NodeSelectorTerm],
+        node_selector_requirements_to_remove: typing.List[
+            client.V1NodeSelectorRequirement
+        ],
+    ) -> typing.List[client.V1NodeSelectorTerm]:
+        """
+        Goes over each expression in all the terms provided and removes the expressions if it matches
+        one of the requirements provided to remove
+
+        :return: New list of terms without the provided node selector requirements
+        """
+        new_node_selector_terms: typing.List[client.V1NodeSelectorTerm] = []
+        # for all term expressions on function spec
+        for term in node_selector_terms:
+            new_node_selector_requirements: typing.List[
+                client.V1NodeSelectorRequirement
+            ] = []
+            for expression in term.match_expressions:
+                new_node_selector_requirements.append(expression)
+                # go over each requirement and check if matches the current expression
+                for node_selector_requirement in node_selector_requirements_to_remove:
+                    if node_selector_requirement == expression:
+                        # remove from new node selector requirements list
+                        new_node_selector_requirements.pop()
+                        # no need to keep going over the list provided for the current expression
+                        break
```

Comment:
I don't like the idea of appending and then popping, much more clean (IMO) to append only when needed (after the loop) and skip (`continue`) when not